rust/worker/src/execution/orchestration/compact.rs

```diff
@@ -679,7 +681,38 @@ impl Handler<TaskResult<MaterializeLogOutput, MaterializeLogOperatorError>>
             }
         };
 
-        self.write(materialized_result, ctx.receiver(), ctx).await;
+        if materialized_result.get_materialized_records().is_empty() {
+            // There is no delta after materialization, so there is no need to write
+            self.num_write_tasks -= 1;
+            if self.num_write_tasks == 0 {
+                if self.dirty {
+                    let writer_res = self.get_segment_writers().await;
+                    let (record_segment_writer, hnsw_segment_writer, metadata_segment_writer) =
+                        match writer_res {
+                            Ok(writers) => writers,
+                            Err(e) => {
+                                tracing::error!("Error creating writers for compaction {:?}", e);
+                                terminate_with_error(self.result_channel.take(), e, ctx);
+                                return;
+                            }
+                        };
+                    self.flush_s3(
+                        record_segment_writer,
+                        hnsw_segment_writer,
+                        metadata_segment_writer,
+                        ctx.receiver(),
+                    )
+                    .await;
+                } else {
+                    // There is no delta across all partition, so there is no need to flush
+                    self.register(self.pulled_log_offset.unwrap(), [].into(), ctx.receiver())
```

Comment:
I believe this is could be a catastrophic but if we had not built register collections in a certain way. Did you audit the backing code? It relies on a subtle property that If the compaction had files in its flush info, it will keep them instead of blowing them away.

I think if we want to rely on this assumption _we must_ add tests for it

https://github.com/chroma-core/chroma/blob/c9498b4d9101326bbe7c3b84dfe476dcd272beec/go/pkg/sysdb/metastore/db/dao/segment.go#L216